node/pkg/watchers/solana/shim.go

```diff
@@ -0,0 +1,362 @@
+// - Publishes the observation.
+func (s *SolanaWatcher) shimProcessInnerInstruction(
+	logger *zap.Logger,
+	whProgramIndex uint16,
+	shimProgramIndex uint16,
+	tx *solana.Transaction,
+	innerInstructions []solana.CompiledInstruction,
+	outerIdx int,
+	startIdx int,
+	alreadyProcessed ShimAlreadyProcessed,
+	isReobservation bool,
+) (bool, error) {
+	// See if this is a PostMessage event from the shim contract. If so, parse it. If not, bail out now.
+	postMessage, err := shimParsePostMessage(s.shimPostMessageDiscriminator, innerInstructions[startIdx].Data)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse inner shim post message instruction %d, %d: %w", outerIdx, startIdx, err)
+	}
+
+	if postMessage == nil {
+		return false, nil
+	}
+
+	alreadyProcessed.add(outerIdx, startIdx)
+
+	err = s.shimProcessRest(logger, whProgramIndex, shimProgramIndex, tx, innerInstructions, outerIdx, startIdx+1, postMessage, alreadyProcessed, isReobservation)
+	if err != nil {
+		return false, fmt.Errorf("failed to process inner instructions for top-level shim instruction %d: %w", outerIdx, err)
+	}
+
+	return true, nil
+}
+
+// shimProcessRest performs the processing of the inner instructions that is common to both the direct and integrator case. It looks for the PostMessage from
+// the core and the MessageEvent from the shim. Note that the startIdx parameter tells us where to start looking for these events. In the direct case, this
+// will be zero. In the integrator case, it is one after the shim PostMessage event.
+func (s *SolanaWatcher) shimProcessRest(
+	logger *zap.Logger,
+	whProgramIndex uint16,
+	shimProgramIndex uint16,
+	tx *solana.Transaction,
+	innerInstructions []solana.CompiledInstruction,
+	outerIdx int,
+	startIdx int,
+	postMessage *ShimPostMessageData,
+	alreadyProcessed ShimAlreadyProcessed,
+	isReobservation bool,
+) error {
+	// Loop through the inner instructions after the shim PostMessage and do the following:
+	// 1) Find the core event and verify it is unreliable with an empty payload.
+	// 2) Find the shim MessageEvent to get the rest of the fields we need for the observation.
+	// 3) Verify that the shim MessageEvent comes after the core event.
+
+	var verifiedCoreEvent bool
+	var messageEvent *ShimMessageEventData
+	var err error
+	coreEventFound := false
+	for idx := startIdx; idx < len(innerInstructions); idx++ {
+		inst := innerInstructions[idx]
+		if inst.ProgramIDIndex == whProgramIndex {
+			if verifiedCoreEvent, err = shimVerifyCoreMessage(inst.Data); err != nil {
+				return fmt.Errorf("failed to verify inner core instruction for shim instruction %d: %w", outerIdx, err)
+			}
+			alreadyProcessed.add(outerIdx, idx)
+			coreEventFound = true
+		} else if inst.ProgramIDIndex == shimProgramIndex {
+			if !coreEventFound {
+				return fmt.Errorf("detected an inner shim message event instruction before the core event for top-level shim instruction %d: %w", outerIdx, err)
```

Comment:
I could go either way on the errors in this function referring to the shim post message instruction as "top-level"